host.py

```diff
@@ -240,18 +241,28 @@ def ssh_connect(self, username: str, password: Optional[str] = None, rsa_path: s
 
     def ssh_connect_looped(self, logins: list[Login]) -> None:
         if len(logins) == 0:
-            raise Exception("No usuable logins found")
-        while True:
-            for e in logins:
+            raise Exception("No usable logins found")
+
+        login_details = ", ".join([str(login.vars()) for login in logins])
+        logger.info(f"Attempting SSH connections on {self._hostname} with logins: {login_details}")
+
+        for _ in range(360):
+            for login in logins:
                 try:
-                    self._host = e.login()
+                    self._host = login.login()
+                    logger.info(f"Login successful on {self._hostname}")
                     return
                 except ssh_exception.AuthenticationException as e:
-                    logger.info(type(e))
-                    raise e
-                except Exception as e:
-                    logger.info(type(e))
+                    logger.error(f"{type(e).__name__} - {str(e)} for login {login.vars()} on host {self._hostname}")
                     time.sleep(10)
+                except (ssh_exception.NoValidConnectionsError, ssh_exception.SSHException, socket.error, socket.timeout) as e:
+                    logger.error(f"Connection issue for login {login.vars()} on host {self._hostname}: {type(e).__name__} - {str(e)}")
+                    time.sleep(10)
+                except Exception as e:
+                    logger.exception(f"SSH connect, login {login.vars()} user {login._username} on host {self._host}: {type(e).__name__} - {str(e)}")
+                    raise e
+        else:
+            raise ConnectionError(f"Failed to establish an SSH connection to {self._hostname}")
```

Comment:
the change to `ssh_connect_looped()` overlaps with commit "host: fix ssh_connect_looped() to not fail after first authentication error" from #133.

It should be consolidated, to take the best of both.

Reply:
That's a good point, I'll consolidate them

Reply:
I took the `timeout` parameter from your PR, and added an `AutoLogin` to handle automatic logins instead of making password optional. If there's anything I missed please let me know.

I didn't add `first_try`, as Paramiko returns an `AuthenticationException` if authentication fails. I think we can continue looping if it fails.

I also think its cleaner to have `timeout` be an `int` instead of a `float` and use a `range` loop, but if its more idiomatic to have it as a float we can do that.

Reply:
> and added an AutoLogin to handle automatic logins instead of making password optional. If there's anything I missed please let me know.

Ah. Does this replace "host: always add a PasswordLogin for ssh-agent" commit from #133? Nice.

But there is now a warning logged:
```bash
$ python -c 'import host; h = host.RemoteHost("127.0.0.1"); h.ssh_connect("root"); print(h.run("whoami").out)'
2024-04-16 16:57:58 INFO: waiting for '127.0.0.1' to respond to ping
2024-04-16 16:57:58 INFO: 127.0.0.1 up, connecting with root
2024-04-16 16:57:58 INFO: Attempting SSH connections on 127.0.0.1 with logins: {'_username': 'root', '_hostname': '127.0.0.1', 'host': <paramiko.client.SSHClient object at 0x7f1f3a8bb260>}
2024-04-16 16:57:58 INFO: Logging into 127.0.0.1 with paramiko auto key discovery & ssh-agent
/data/src/cluster-deployment-automation/ocp-venv/lib64/python3.12/site-packages/paramiko/client.py:852: UserWarning: Unknown ssh-ed25519 host key for 127.0.0.1: b'7b9652185f0b32d5f059088ae703aa76'
  warnings.warn(
2024-04-16 16:57:58 INFO: Login successful on 127.0.0.1
root

```

Reply:
The warning is because I added [     self.host.set_missing_host_key_policy(paramiko.WarningPolicy())](https://github.com/bn222/cluster-deployment-automation/pull/162/files#diff-ccff53274e18cb1bf0c3ca0194bf14ee709a8b2a35a74e0ebce33feee69b9c30R45)

I thought it made sense to warn about unknown host keys, but we can change it back to `AutoAddPolicy` if its confusing

Reply:
I think, if we want to warn, then we would need at least load the already known hosts from `~/.ssh/known_hosts`. Don't we?

See https://docs.paramiko.org/en/latest/api/hostkeys.html .

But this seems not a useful warning to me in the context of CDA...

Reply:
It is a good point that we need to load the hosts. I added a line for [self.host.load_system_host_keys()](https://github.com/bn222/cluster-deployment-automation/pull/162/files#diff-ccff53274e18cb1bf0c3ca0194bf14ee709a8b2a35a74e0ebce33feee69b9c30R45)

Reply:
for some reason, it no longer works with `load_system_host_keys()`:

`bash
$ python -c 'import host; h = host.RemoteHost("127.0.0.1"); h.ssh_connect("root"); print(h.run("whoami").out)'
2024-04-16 19:23:00 INFO: waiting for '127.0.0.1' to respond to ping
2024-04-16 19:23:00 INFO: 127.0.0.1 up, connecting with root
2024-04-16 19:23:00 INFO: Attempting SSH connections on 127.0.0.1 with logins: {'_username': 'root', '_hostname': '127.0.0.1', 'host': <paramiko.client.SSHClient object at 0x7f8a52b5c470>}
2024-04-16 19:23:00 INFO: Logging into 127.0.0.1 with paramiko auto key discovery & ssh-agent
2024-04-16 19:23:00 INFO: BadHostKeyException - Host key for server '127.0.0.1' does not match: got 'AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBEf2e0VRUq6K3ZASIxJu9sl27r+0MCxTWDDP/H9TaavHK1u9qDLZWiUXr6IyS1I/SgAk1aPFBxjUxWmyecs+LaQ=', expected 'AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBJNh+bjKUI5kaXD9HwOmqY+66wWMs2NXkEhMh12l+EFI3V1jJJHJOhS6UqKg9FBQTuvp5n92/CfCV8T3467AiY8=' for login {'_username': 'root', '_hostname': '127.0.0.1', 'host': <paramiko.client.SSHClient object at 0x7f8a52b5c470>} on host 127.0.0.1
2024-04-16 19:23:10 INFO: Logging into 127.0.0.1 with paramiko auto key discovery & ssh-agent
2024-04-16 19:23:20 INFO: Logging into 127.0.0.1 with paramiko auto key discovery & ssh-agent
2024-04-16 19:23:30 INFO: Logging into 127.0.0.1 with paramiko auto key discovery & ssh-agent
...
```